### PR TITLE
Feature/angular redux and ngrx

### PR DESCRIPTION
**Pull request by AnatoliRub**
**_Task scope:_**
 - [x] Admin page is implemented 
 
- [x] The Admin page functionality with necessary rules is implemented

- [x]  NgRX library is used and storage is created

- [x] Custom cards are saved in the store

- [x] Saving data received from the YouTube API is implemented using NgRX library

### CODE REVIEW

File: youtube-client/src/app/app.module.ts [hunk 1]
```diff
@@ -1,16 +1,34 @@
...
+    StoreModule.forRoot(reducers, {
+      metaReducers,
+      runtimeChecks: {
+        strictStateImmutability: true,
+        strictActionImmutability: true,
+        strictActionSerializability: true,
+        strictStateSerializability: true,
+      },
+    }),
+    StoreDevtoolsModule.instrument({ maxAge: 25, logOnly: environment.production }),
```
Comment (StanevichIvan) [inline youtube-client/src/app/app.module.ts:29]: Do we want to have Devtools in prod mode?

File: youtube-client/src/app/shared/components/button/button.component.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+  @Output() clickHandler: EventEmitter<void> = new EventEmitter<void>();
+
+  setStyles(): {
+    icon: boolean;
+    text: boolean;
+    circle: boolean;
+    link: boolean;
+  } {
+    return {
```
Comment (StanevichIvan) [inline youtube-client/src/app/shared/components/button/button.component.ts:26]: might this logic could be placed in template?

File: youtube-client/src/app/technical/pages/error-page/error-page.component.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  constructor(private readonly route: ActivatedRoute) {}
+
+  ngOnInit(): void {
+    this.subscription = this.route.queryParams.subscribe((val) => {
+      this.adminError = val['admin'];
+    });
+  }
+
+  ngOnDestroy(): void {
+    this.subscription?.unsubscribe();
```
Comment (StanevichIvan) [inline youtube-client/src/app/technical/pages/error-page/error-page.component.ts:24]: let's create disposable abstract class and inherit it

**Review by iamsixthousand — APPROVED** [review]

File: youtube-client/src/app/technical/pages/error-page/error-page.component.ts (hunk 3 above)
Comment (AnatoliRub) [reply youtube-client/src/app/technical/pages/error-page/error-page.component.ts]: 907544d

File: youtube-client/src/app/shared/components/button/button.component.ts (hunk 2 above)
Comment (AnatoliRub) [reply youtube-client/src/app/shared/components/button/button.component.ts]: f4be2fc

File: youtube-client/src/app/app.module.ts (hunk 1 above)
Comment (AnatoliRub) [reply youtube-client/src/app/app.module.ts]: b9856d7

File: youtube-client/src/app/auth/ngrx/actions/auth.actions.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import { createAction, props } from '@ngrx/store';
+import { UserData } from 'src/app/admin/types/user-data.type';
+
+export const login = createAction('[Login form] login', props<{ user: UserData }>());
```
[the remark this answers is not in the document]
Comment (AnatoliRub) [reply youtube-client/src/app/auth/ngrx/actions/auth.actions.ts:4]: this piece of code contains the type if you hover in the code editor
"const login: ActionCreator<"[Login form] login", (props: {user: UserData;}) => {user: UserData;} & TypedAction<"[Login form] login">>"

File: youtube-client/src/app/auth/ngrx/effects/auth.effects.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { tap } from 'rxjs';
+import { RoutesPath } from 'src/app/routes.enum';
+import { AuthActions } from '../actions/auth-action.types';
+
+@Injectable()
+export class AuthEffects {
+  login$ = createEffect(
```
[the remark this answers is not in the document]
Comment (AnatoliRub) [reply youtube-client/src/app/auth/ngrx/effects/auth.effects.ts:10]: this piece of code contains the type if you hover in the code editor
"login$: Observable<{user: UserData;} & TypedAction<"[Login form] login">> & CreateEffectMetadata"

File: youtube-client/src/app/core/ngrx/selectors/search.selectors.ts [hunk 6]
```diff
@@ -0,0 +1,14 @@
+import { createFeatureSelector, createSelector } from '@ngrx/store';
+import { SearchState } from '../reducer/search.reducer';
+
+export const selectSerchFeatureSelector = createFeatureSelector<SearchState>('search');
+
+export const selectSearchFilter = createSelector(
```
[the remark this answers is not in the document]
Comment (AnatoliRub) [reply youtube-client/src/app/core/ngrx/selectors/search.selectors.ts:6]: this piece of code contains the type if you hover in the code editor 
"const selectSearchFilter: MemoizedSelector<object, IFilterData, DefaultProjectorFn<IFilterData>>"

File: youtube-client/src/app/youtube/ngrx/selectors/posts.selectors.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { createFeatureSelector, createSelector } from '@ngrx/store';
+import { PostsState } from '../reducers/post.reducer';
+
+import * as fromPosts from '../reducers/post.reducer';
+
+export const selectPostsFeatureSelector = createFeatureSelector<PostsState>('posts');
```
[the remark this answers is not in the document]
Comment (AnatoliRub) [reply youtube-client/src/app/youtube/ngrx/selectors/posts.selectors.ts:6]: this piece of code contains the type if you hover in the code editor 
"const selectPostsFeatureSelector: MemoizedSelector<object, PostsState, DefaultProjectorFn<PostsState>>"

**Review by StanevichIvan — APPROVED** [review]

File: youtube-client/src/app/youtube/pipes/format-count.pipe.spec.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import { FormatCountPipe } from './format-count.pipe';
+
+describe('FormatCountPipe', () => {
+  let pipe: FormatCountPipe;
```
[the remark this answers is not in the document]
Comment (AnatoliRub) [reply youtube-client/src/app/youtube/pipes/format-count.pipe.spec.ts:4]: It's my first test which covered all branches.^.^, Yoooohooo)))

File: youtube-client/src/app/technical/pages/error-page/error-page.component.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+export class ErrorPageComponent extends Disposable implements OnInit {
+  adminError?: boolean;
+
+  constructor(private readonly route: ActivatedRoute) {
+    super();
+  }
+
+  ngOnInit(): void {
+    this.route.queryParams.pipe(takeUntil(this.disposed$)).subscribe((val) => {
+      this.adminError = val['admin'];
```
[the remark this answers is not in the document]
Comment (AnatoliRub) [reply youtube-client/src/app/technical/pages/error-page/error-page.component.ts]: It's cool idea.:)
bfe210b

**Review by StanevichIvan — APPROVED** [review]

**Pull request merged**